src/structures/paging/mapper/mod.rs

```diff
@@ -275,11 +318,172 @@ pub trait Mapper<S: PageSize> {
+    where
+        Self: Sized,
+        A: FrameAllocator<Size4KiB> + FrameAllocator<S> + ?Sized,
+    {
+        pages
+            .try_for_each(|page| {
+                let frame = frame_allocator
+                    .allocate_frame()
+                    .ok_or((MapToError::FrameAllocationFailed, page))?;
+
+                unsafe {
+                    // SAFETY: frame was just freshly allocated and therefore can't cause aliasing issues
+                    self.map_to_with_table_flags(
+                        page,
+                        frame,
+                        flags,
+                        parent_table_flags,
+                        frame_allocator,
+                    )
+                }
+                .map(|_| ())
+                .map_err(|e| (e, page))
+            })
+            .map(|_| MapperFlushRange::new(pages))
+            .map_err(|(e, page)| {
+                (
+                    e,
+                    MapperFlushRange::new(PageRange {
+                        start: pages.start,
+                        end: page,
+                    }),
+                )
+            })
+    }
+
+    /// Maps frames from the allocator to the given range of virtual pages.
+    ///
+    /// ## Errors
+    ///
+    /// If an error occurs half-way through a [`MapperFlushRange<S>`] is returned that contains the frames that were successfully mapped.
+    #[inline]
+    fn map_range<A>(
+        &mut self,
+        pages: PageRange<S>,
+        flags: PageTableFlags,
+        frame_allocator: &mut A,
+    ) -> Result<MapperFlushRange<S>, (MapToError<S>, MapperFlushRange<S>)>
+    where
+        Self: Sized,
+        A: FrameAllocator<Size4KiB> + FrameAllocator<S> + ?Sized,
+    {
+        let parent_table_flags = flags
+            & (PageTableFlags::PRESENT
+                | PageTableFlags::WRITABLE
+                | PageTableFlags::USER_ACCESSIBLE);
+        self.map_range_with_table_flags(pages, flags, parent_table_flags, frame_allocator)
+    }
+
     /// Removes a mapping from the page table and returns the frame that used to be mapped.
     ///
     /// Note that no page tables or pages are deallocated.
     fn unmap(&mut self, page: Page<S>) -> Result<(PhysFrame<S>, MapperFlush<S>), UnmapError>;
 
+    /// Removes a range of mapping from the page table and deallocate the frames that used to be mapped.
+    ///
+    /// Note that no page tables or pages are deallocated.
+    ///
+    /// ## Errors
+    /// If an error occurs half-way through a [`MapperFlushRange<S>`] is returned that contains the pages that were successfully unmapped.
+    ///
+    /// ## Safety
+    /// The caller has to guarantee that it's safe to deallocate frames:
+    /// All unmapped frames must only be only in this page table.
+    unsafe fn unmap_range<D>(
+        &mut self,
+        pages: PageRange<S>,
+        deallocator: &mut D,
+    ) -> Result<MapperFlushRange<S>, (UnmapError, MapperFlushRange<S>)>
```

Comment:
I like the idea, but I'm unsure on how to implement it.
1. When do you propose we unmap the pages? When the method is called or when the returned iterator is used?
2. What should the return type of this method be? Should all implementations return the same iterator or should we add an associated type for the iterator (this would make the trait non object safe)?

Reply:
> 1. When do you propose we unmap the pages? When the method is called or when the returned iterator is used?

Good point! Given that we cannot allocate, I think only the second approach is possible (when the iterator is used).

> 2. What should the return type of this method be? Should all implementations return the same iterator or should we add an associated type for the iterator (this would make the trait non object safe)?

Hmm, I forgot that `impl Trait` return types are not supported yet for traits. This makes things much more complicated...

Let's keep the implementation as it is right now. It's better to have the small inconsistency between `unmap` and `unmap_range` than to make the API more complex.